### PR TITLE
Refactor/tag creator

### LINKED ISSUE #584: Creator - generates tags with invalid name
When the displayName of a tag declared at the service level contains spaces or punctuation, this generates an invalid tags template:

config file:
```
tags: 
  - displayName: Internal Payroll - International
```

produced template:

```
    {
      "properties": {
        "displayName": "Internal Payroll - International"
      },
      "name": "[concat(parameters('ApimServiceName'), '/Internal Payroll - International')]",
      "type": "Microsoft.ApiManagement/service/tags",
      "apiVersion": "2021-01-01-preview",
      "dependsOn": []
    }

```
This yields an error message on deploying:

`2021-10-14T12:03:51.9706948Z ##[error]ValidationError: Invalid identifier Internal Payroll - International specified.  Entity Identifier can contain only letters, numbers, and hyphens. The first and last character must be a letter or a number `

There is no documented way to specify any properties of tags other than their displayName, which appears to be carried through into the internal resource name as though it is a valid name, which it may not be. When created in a live APIM instance, the displayname is as stated but the internal resource name is internal-payroll-international. There does not appear to be a way to make the Creator do this conversion.

### LINKED ISSUE #730: refactor `TagCreator`
Refactor TagCreator 
Validate/add tests
Code style

### PR DESCRIPTION
**Pull request by f-alizada**
Refactor and change behaviour of the tag creator class.
Display name is converted and sanitized to resource name for tags. 
Run the tag creator in case there are no tags presented in main configs but only in ApiConfig
Tests introduced for changes
Error messages class created for proper unit testing without duplication of messages. 

Closes #730 
Fixes #584

### CODE REVIEW

File: src/ArmTemplates/Commands/Executors/CreatorExecutor.cs [hunk 1]
```diff
@@ -205,7 +206,7 @@ public async Task ExecuteGenerationBasedOnConfiguration()
 
             this.logger.LogInformation("Creating tag template");
             this.logger.LogInformation("------------------------------------------");
-            var tagTemplate = this.creatorParameters.Tags != null ? this.tagTemplateCreator.CreateTagTemplate(this.creatorParameters) : null;
+            var tagTemplate = !this.creatorParameters.Tags.IsNullOrEmpty() || this.creatorParameters.Apis.Any(x => !x.Tags.IsNullOrEmpty()) ? this.tagTemplateCreator.CreateTagTemplate(this.creatorParameters) : null;
```
Comment (DeagleGross) [inline src/ArmTemplates/Commands/Executors/CreatorExecutor.cs:209]: Let's move all the section of `tagTemplate` creation to a separate method, as it is done it `ExtractorExecutor` - there in the `GenerateTemplates` method we call others methods, that encapsulate logic of creating that or another template.

I would go with this approach for several reasons:
- easy testing of some occasional use-cases - we do not need to run whole `ExecuteBasedOnConfiguration` only to check `tag-template`
- easy debug of failures - stack trace tells you directly what method failed. Currently you will have to find "that specific line" of failure in `ExecuteBasedOnConfiguration` method
- I know it is quite early to think about that, but some users requested `resource-kit` as a NuGet. So `ExtractorExecutor` has the API for that:
```
var extractorExecutor = new ExtractorExecutor(...);
await extractorExecutor.GenerateTagTemplates();
```
But in current implementation you do not have a possibility to do this. If it is not that hard to (and I think it is not), then it is better to already move logic to separate methods. Also don't forget to remove that dummy `-------` log :D

File: src/ArmTemplates/Common/Exceptions/DuplicateTagResourceNameException.cs [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+//  Licensed under the MIT License.
+// --------------------------------------------------------------------------
+
+using System;
+
+namespace Microsoft.Azure.Management.ApiManagement.ArmTemplates.Common.Exceptions
+{
+    public class DuplicateTagResourceNameException : Exception
+    {
+        public DuplicateTagResourceNameException(string message) : base(message)
```
Comment (DeagleGross) [inline src/ArmTemplates/Common/Exceptions/DuplicateTagResourceNameException.cs:12]: you can encapsulate logic of `string.Format` here. Like this:
```suggestion
         const string DuplicateTagResourceNameErrorMessageFormat = "Duplicate tag resource name found during sanitizing the display name. Please consider renaming tags: {0}, {1}. Both resulted resource name to be equal to: {2}";
 
        public DuplicateTagResourceNameException(string existingValue, string tagName, string resourceName) 
            : base(string.Format(ErrorMessages.DuplicateTagResourceNameErrorMessage, existingValue, tagName, resourceName))
```

My points for that implementation:
- I do not think `DuplicateTagResourceNameErrorMessageFormat` is used in any other exception, so we can easily hide it in corresponding exception
- usage of this exception becomes much more easy to understand - you tell developer explicitly what to pass to show in exception
```
throw new DuplicateTagResourceNameException(existingValue, tagName, resourceName);
```

File: src/ArmTemplates/Common/Extensions/ResourceNamingHelper.cs [hunk 3]
```diff
@@ -0,0 +1,26 @@
+﻿// --------------------------------------------------------------------------
+//  Copyright (c) Microsoft Corporation. All rights reserved.
+//  Licensed under the MIT License.
+// --------------------------------------------------------------------------
+
+using System.Text.RegularExpressions;
+
+namespace Microsoft.Azure.Management.ApiManagement.ArmTemplates.Common.Extensions
+{
+    public static class ResourceNamingHelper
```
Comment (DeagleGross) [inline src/ArmTemplates/Common/Extensions/ResourceNamingHelper.cs:10]: Duplicate with [ParameterNamingHelper](https://github.com/Azure/azure-api-management-devops-resource-kit/blob/main/src/ArmTemplates/Common/Extensions/ParameterNamingHelper.cs). Or do I miss anything?

File: src/ArmTemplates/Creator/TemplateCreators/TagTemplateCreator.cs [hunk 4]
```diff
@@ -24,51 +26,64 @@ public TagTemplateCreator(ITemplateBuilder templateBuilder)
 
         public Template CreateTagTemplate(CreatorParameters creatorConfig)
```
Comment (DeagleGross) [inline src/ArmTemplates/Creator/TemplateCreators/TagTemplateCreator.cs:27]: Can we return `Template<TTemplateResource>` here? I am not sure it is very important right now, but it allows us to reuse already handled resources by future templateCreators. The same is done in extractor

What do you think?

File: src/ArmTemplates/Creator/TemplateCreators/TagTemplateCreator.cs [hunk 5]
```diff
@@ -24,51 +26,64 @@ public TagTemplateCreator(ITemplateBuilder templateBuilder)
...
         {
-            // create empty template
-            Template tagTemplate = this.templateBuilder.GenerateEmptyTemplate().Build();
+            var tagTemplate = this.templateBuilder.GenerateTemplateWithApimServiceNameProperty().Build();
+            var tagsDictionary = new Dictionary<string, string>();
 
-            // add parameters
-            tagTemplate.Parameters = new Dictionary<string, TemplateParameterProperties>
+            void AddTagNameToDictionary(string tagName, Dictionary<string, string> tagsDictionary)
```
Comment (DeagleGross) [inline src/ArmTemplates/Creator/TemplateCreators/TagTemplateCreator.cs:32]: 1. what is the reason to make method inside a method? It is a little bit frustrating for me at first sight
2. Even if you want to use inner-method approach, then there is no reason for parameter `Dictionary<string, string> tagsDictionary` - it can reuse the `tagsDictionary` itself

File: src/ArmTemplates/Creator/TemplateCreators/TagTemplateCreator.cs [hunk 6]
```diff
@@ -24,51 +26,64 @@ public TagTemplateCreator(ITemplateBuilder templateBuilder)
...
             {
-                foreach (ApiConfig api in apis)
+                foreach (var api in creatorConfig.Apis)
                 {
-                    if (api.Tags != null)
+                    if (!api.Tags.IsNullOrEmpty())
                     {
-                        string[] apiTags = api.Tags.Split(", ");
-                        foreach (string apiTag in apiTags)
+                        var apiTags = api.Tags.Split(", ");
```
Comment (DeagleGross) [inline src/ArmTemplates/Creator/TemplateCreators/TagTemplateCreator.cs:56]: Please, add a corner case test for this code like
`apiTags = mytag1,mytag2, mytag3, `

Comment (DeagleGross) [reply src/ArmTemplates/Creator/TemplateCreators/TagTemplateCreator.cs:56]: I guess 3 tags have to be parsed - myTag1, myTag2 and myTag3

File: src/ArmTemplates/Creator/TemplateCreators/TagTemplateCreator.cs (hunk 5 above)
Comment (DeagleGross) [reply src/ArmTemplates/Creator/TemplateCreators/TagTemplateCreator.cs:32]: We can also build a collection of resources here - just add that `TagDictionary` to `TagResources` and we will do 2 things:
- cache tags dictionary for future, if needed
- build resources for current template built right away

File: tests/ArmTemplates.Tests/Common/ExtensionsTest.cs [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+using Xunit;
+using Microsoft.Azure.Management.ApiManagement.ArmTemplates.Common.Extensions;
+using FluentAssertions;
+
+namespace Microsoft.Azure.Management.ApiManagement.ArmTemplates.Tests.Common
+{
+    public class ExtensionsTest
+    {
+        [Fact]
+        public void GenerateValidResourceNameFromDisplayName_ShouldReturnNull_GivenEmptyString()
```
Comment (DeagleGross) [inline tests/ArmTemplates.Tests/Common/ExtensionsTest.cs:15]: If you want to use `ParameterNamingHelper` these tests will be helpful for that class!

File: tests/ArmTemplates.Tests/Creator/TemplateCreatorTests/TagTemplateCreatorTests.cs [hunk 8]
```diff
@@ -10,28 +10,150 @@
...
+using System.Linq;
+using FluentAssertions;
+using Microsoft.Azure.Management.ApiManagement.ArmTemplates.Common.Exceptions;
+using Microsoft.Azure.Management.ApiManagement.ArmTemplates.Common.Extensions;
 
 namespace Microsoft.Azure.Management.ApiManagement.ArmTemplates.Tests.Creator.TemplateCreatorTests
 {
     public class TagTemplateCreatorTests
     {
+
```
Comment (DeagleGross) [inline tests/ArmTemplates.Tests/Creator/TemplateCreatorTests/TagTemplateCreatorTests.cs:22]: indent

File: tests/ArmTemplates.Tests/Common/ExtensionsTest.cs [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+//  Licensed under the MIT License.
+// --------------------------------------------------------------------------
+
+using Xunit;
+using Microsoft.Azure.Management.ApiManagement.ArmTemplates.Common.Extensions;
+using FluentAssertions;
+
+namespace Microsoft.Azure.Management.ApiManagement.ArmTemplates.Tests.Common
+{
+    public class ExtensionsTest
```
Comment (DeagleGross) [inline tests/ArmTemplates.Tests/Common/ExtensionsTest.cs:12]: We categorize tests with `trait` - [Organizing tests with Traits in XUnit](http://www.brendanconnolly.net/organizing-tests-with-xunit-traits/) (take a look at extractor scenarios tests). Personally I never use them, but maybe at some time it will be usefull.

We can either **remove them completely** or **add them here also**. What do you think?

Comment (DeagleGross) [reply tests/ArmTemplates.Tests/Common/ExtensionsTest.cs:12]: This applies to other tests as well

File: tests/ArmTemplates.Tests/Creator/TemplateCreatorTests/TagTemplateCreatorTests.cs [hunk 10]
```diff
@@ -10,28 +10,150 @@
...
             //assert
-            Assert.Equal($"[concat(parameters('{ParameterNames.ApimServiceName}'), '/{tag.DisplayName}')]", tagTemplateResource.Name);
+            tagTemplate.Parameters.Count().Should().Be(1);
+            tagTemplate.Parameters.ContainsKey(ParameterNames.ApimServiceName).Should().BeTrue();
+
+            tagTemplate.Resources.Count().Should().Be(5);
+            
+            foreach (TagTemplateResource tag in tagTemplate.Resources)
+            {
+                var resourceTagName = tag.Name.Split("/")[1].Split("'")[0];
```
Comment (DeagleGross) [inline tests/ArmTemplates.Tests/Creator/TemplateCreatorTests/TagTemplateCreatorTests.cs:92]: If I see this, I am all for saving original name of tag somewhere in resource. Such logic in tests is very error-prone IMO.

File: tests/ArmTemplates.Tests/Creator/TemplateCreatorTests/TagTemplateCreatorTests.cs [hunk 11]
```diff
@@ -10,28 +10,150 @@
...
+            {
+                var tag = new TagProperties
+                {
+                    DisplayName = tagName
+                };
+                creatorConfig.Tags.Add(tag);
+            }
+
+            //act & assert
+            var exception = Assert.Throws<DuplicateTagResourceNameException>(() => tagTemplateCreator.CreateTagTemplate(creatorConfig));
```
Comment (DeagleGross) [inline tests/ArmTemplates.Tests/Creator/TemplateCreatorTests/TagTemplateCreatorTests.cs:155]: we have `FluentAssertions` in `resource-kit`. This allows to right fancy `.Should()` methods to assert something.
And they have an implementation for checking exceptions: [ShouldThrow()](https://fluentassertions.com/exceptions/)

File: src/ArmTemplates/Common/Exceptions/DuplicateTagResourceNameException.cs (hunk 2 above)
Comment (f-alizada) [reply src/ArmTemplates/Common/Exceptions/DuplicateTagResourceNameException.cs:12]: I was thinking about this approach but decided to go with this one to have all error messages in one place but not in separate classes per error. What do you think?

File: src/ArmTemplates/Common/Extensions/ResourceNamingHelper.cs (hunk 3 above)
Comment (f-alizada) [reply src/ArmTemplates/Common/Extensions/ResourceNamingHelper.cs:10]: I didn't want to change the already existing ParameterNamingHelper since related to the parameter naming, and for the resourceNaming helper it has different rules for example it could have in the name '-' char. Also, in the future each of this classes could be extended with resource/parameter naming functionality.

File: src/ArmTemplates/Common/Exceptions/DuplicateTagResourceNameException.cs (hunk 2 above)
Comment (f-alizada) [reply src/ArmTemplates/Common/Exceptions/DuplicateTagResourceNameException.cs]: As was discussed updated the exceptions classes with custom constructors per each introduced exception

File: src/ArmTemplates/Common/Extensions/ResourceNamingHelper.cs (hunk 3 above)
Comment (f-alizada) [reply src/ArmTemplates/Common/Extensions/ResourceNamingHelper.cs]: Combined the Parameter and Resource naming helpers in one NamingHelper.

File: src/ArmTemplates/Creator/TemplateCreators/TagTemplateCreator.cs (hunk 6 above)
Comment (f-alizada) [reply src/ArmTemplates/Creator/TemplateCreators/TagTemplateCreator.cs]: `apiTags = mytag1,mytag2, mytag3,` This example in PR changes will raise an Exception with EmptyResourceName message

File: tests/ArmTemplates.Tests/Creator/TemplateCreatorTests/TagTemplateCreatorTests.cs (hunk 8 above)
Comment (f-alizada) [reply tests/ArmTemplates.Tests/Creator/TemplateCreatorTests/TagTemplateCreatorTests.cs]: updated thank you

File: tests/ArmTemplates.Tests/Common/ExtensionsTest.cs (hunk 9 above)
Comment (f-alizada) [reply tests/ArmTemplates.Tests/Common/ExtensionsTest.cs:13]: Agree added to introduced and modified tests

File: tests/ArmTemplates.Tests/Creator/TemplateCreatorTests/TagTemplateCreatorTests.cs (hunk 10 above)
Comment (f-alizada) [reply tests/ArmTemplates.Tests/Creator/TemplateCreatorTests/TagTemplateCreatorTests.cs]: Updated the method of checking with relying on common method GenerateParametrizedResourceName

File: tests/ArmTemplates.Tests/Creator/TemplateCreatorTests/TagTemplateCreatorTests.cs (hunk 11 above)
Comment (f-alizada) [reply tests/ArmTemplates.Tests/Creator/TemplateCreatorTests/TagTemplateCreatorTests.cs]: Thank you, updated

**Review by DeagleGross — COMMENTED** [review]
Added a couple of little code comments, but overall this is great start of creator refactoring!

File: src/ArmTemplates/Commands/Executors/CreatorExecutor.cs [hunk 12]
```diff
@@ -242,44 +241,57 @@ public async Task ExecuteGenerationBasedOnConfiguration()
...
             // write parameters to outputLocation
-            FileWriter.WriteJSONToFile(templateParameters, string.Concat(this.creatorParameters.OutputLocation, this.creatorParameters.FileNames.Parameters));
+            FileWriter.WriteJSONToFile(templateParameters, Path.Combine(this.creatorParameters.OutputLocation, this.creatorParameters.FileNames.Parameters));
             this.logger.LogInformation("Templates written to output location");
         }
+
+        public async Task<Template> GenerateTagsTemplate()
+        {
+            this.logger.LogInformation("Creating tag template");
+            var tagTemplate = !this.creatorParameters.Tags.IsNullOrEmpty() || this.creatorParameters.Apis.Any(x => !x.Tags.IsNullOrEmpty()) ? this.tagTemplateCreator.CreateTagTemplate(this.creatorParameters) : null;
```
[the remark this answers is not in the document]
Comment (f-alizada) [reply src/ArmTemplates/Commands/Executors/CreatorExecutor.cs]: Updated :)

File: src/ArmTemplates/Commands/Executors/CreatorExecutor.cs [hunk 13]
```diff
@@ -242,44 +241,57 @@ public async Task ExecuteGenerationBasedOnConfiguration()
...
+
+            if (tagTemplate != null)
```
[the remark this answers is not in the document]
Comment (f-alizada) [reply src/ArmTemplates/Commands/Executors/CreatorExecutor.cs]: Taking into account that previous comment resolved then yes it does not make sense to check for null, before it was :) 
Updated accordingly

File: src/ArmTemplates/Creator/TemplateCreators/TagTemplateCreator.cs [hunk 14]
```diff
@@ -22,53 +24,71 @@ public TagTemplateCreator(ITemplateBuilder templateBuilder)
...
-            Template tagTemplate = this.templateBuilder.GenerateEmptyTemplate().Build();
+            var resourceName = NamingHelper.GenerateValidResourceNameFromDisplayName(tagName);
 
-            // add parameters
-            tagTemplate.Parameters = new Dictionary<string, TemplateParameterProperties>
+            if (string.IsNullOrEmpty(resourceName)) 
             {
-                {ParameterNames.ApimServiceName, new TemplateParameterProperties(){ Type = "string" }}
-            };
+                throw new EmptyResourceNameException(tagName);
```
[the remark this answers is not in the document]
Comment (f-alizada) [reply src/ArmTemplates/Creator/TemplateCreators/TagTemplateCreator.cs:33]: As was discussed, it will be a good approach to have a local validator for the input parameters, so the error could be found before the actual deployment. This kind of validations are helpful for the **creator**

**Review by DeagleGross — APPROVED** [review]
